### PR TITLE
Use create2 when deploying Registry and OrderFactory

### LINKED ISSUE #1121: Use create2 instead of create in registry, factory
Use create2 instead of create in orders registry, orders factory contracts deployment (Deploy.s.sol)

### PR DESCRIPTION
**Pull request by Svetomech**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new `Create2` contract for deploying contracts using the CREATE2 opcode.
	- Enhanced error handling with custom error types for deployment failures.
	- Updated deployment logic for `Registry` and `OrderFactory` contracts to utilize CREATE2.
	- Added new variables in the `justfile` for enhanced configurability of RPC URL and chain ID.
	- Introduced new configuration variables in the `.env.example` file for improved project setup.

- **Bug Fixes**
	- Improved error handling in the `Deploy` contract constructor for invalid addresses.

- **Tests**
	- Added a test contract `Create2Test` to verify the functionality of the `Create2` contract.

### CODE REVIEW

File: solidity/orders/script/Deploy.s.sol [hunk 1]
```diff
@@ -19,24 +23,26 @@ contract Deploy is Script {
...
-            revert InvalidScheduler();
-        }
-        if (factoryOwner == address(0)) {
-            revert InvalidFactory();
-        }
+
+        if (scheduler == address(0)) revert InvalidScheduler();
+        if (factoryOwner == address(0)) revert InvalidFactory();
+
+        registrySalt = keccak256(abi.encodePacked(vm.envOr("REGISTRY_SALT", string("DEFAULT_REGISTRY_SALT"))));
```
Comment (mn13) [inline solidity/orders/script/Deploy.s.sol]: Can you please add this env into .env.example file?

File: solidity/orders/script/Deploy.s.sol [hunk 2]
```diff
@@ -19,24 +23,26 @@ contract Deploy is Script {
...
+        orderFactorySalt = keccak256(abi.encodePacked(vm.envOr("FACTORY_SALT", string("DEFAULT_FACTORY_SALT"))));
```
Comment (mn13) [inline solidity/orders/script/Deploy.s.sol]: And this :)

[the remark this answers is not in the document]
Comment (Svetomech) [reply solidity/orders/script/Deploy.s.sol]: removed envs as they are not needed

File: solidity/orders/script/Deploy.s.sol (hunk 1 above)
Comment (Svetomech) [reply solidity/orders/script/Deploy.s.sol]: removed envs as they are not needed

File: solidity/orders/script/Deploy.s.sol (hunk 2 above)
Comment (Svetomech) [reply solidity/orders/script/Deploy.s.sol]: removed envs as they are not needed

File: solidity/orders/script/Deploy.s.sol [hunk 3]
```diff
@@ -1,43 +1,70 @@
...
+
     function run() external {
         vm.startBroadcast(broadcaster);
 
-        Registry registry;
         if (registryAddress == address(0)) {
-            registry = new Registry();
-        } else {
-            registry = Registry(registryAddress);
+            bytes32 registrySalt = bytes32(abi.encodePacked(broadcaster, hex"00", bytes11(uint88(1001))));
```
[the remark this answers is not in the document]
Comment (Svetomech) [reply solidity/orders/script/Deploy.s.sol]: dont need uniqueness here. Registry address should be deterministic and Factory address is deterministic but dependent on constructor arguments, so no need for a unique salt

File: solidity/orders/src/Create2.sol [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            let ptr := mload(0x40)
+
+            mstore(add(ptr, 0x40), creationCodeHash)
+            mstore(add(ptr, 0x20), salt)
+            mstore(ptr, contractAddress)
+            let start := add(ptr, 0x0b)
+            mstore8(start, 0xff)
+            addr := keccak256(start, 85)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Svetomech) [reply solidity/orders/src/Create2.sol:39]: the code taken is from https://book.getfoundry.sh/tutorials/create2-tutorial and is covered with a test

File: solidity/justfile [hunk 5]
```diff
@@ -1,5 +1,7 @@
 shulgin_mnemonic := "exclude try nephew main caught favorite tone degree lottery device tissue tent ugly mouse pelican gasp lava flush pen river noise remind balcony emerge"
 shulgin := "0x6EA8AC1673402989E7B653AE4E83B54173719C30"
+rpc_url := "http://127.0.0.1:8545"
```
Comment (mn13) [inline solidity/justfile]: `local_rpc_url `

File: solidity/justfile [hunk 6]
```diff
@@ -1,5 +1,7 @@
...
+chain_id := "1337"
```
Comment (mn13) [inline solidity/justfile]: `local_chain_id`

File: solidity/orders/src/Create2.sol [hunk 7]
```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.8.25 <0.9.0;
+
+contract Create2 {
```
Comment (mn13) [inline solidity/orders/src/Create2.sol:5]: Can you pls keep here the link to the source as comment?
`// From https://book.getfoundry.sh/tutorials/create2-tutorial`

**Review by mn13 — CHANGES_REQUESTED** [review]

**Review by mn13 — APPROVED** [review]
LGTM

**Pull request merged**